azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java

```diff
@@ -1,126 +1,201 @@
-        String jobPath = jobDefinitionId;
-        String[] nodesInPath = jobPath.split(":");
-        List<ExecutableNode> nodesToScan = executableFlow.getExecutableNodes();
-        ExecutableNode node = null;
-        for (String nodeId: nodesInPath) {
-            int i=0;
-            for (; i < nodesToScan.size(); i++) {
-                ExecutableNode en = nodesToScan.get(i);
-                if (en.getId().equals(nodeId)) {
-                    node = en;
-                    break;
-                }
-            }
-            if (i >= nodesToScan.size()) {
-                errorMsg = String.format("Job with id %s and path %s wasn't found in execution %d.",
-                    jobDefinitionId, jobPath, execId);
-                logger.error(errorMsg);
-                throw new CloudFlowNotFoundException(errorMsg);
-            }
-            if (node instanceof ExecutableFlowBase) {
-                nodesToScan = ((ExecutableFlowBase) node).getExecutableNodes();
-            } else {
-                // If we find a job before the end of the path this ensures the next element in
-                // the path will not be found in the next iteration. This has no effect with the
-                // last node in the path.
-                nodesToScan = new ArrayList<>();
-            }
-        }
-        logger.info("Found job with path {} in execution {}.", jobPath, execId);
+public class ExecutionServiceImpl implements ExecutionService {
 
-        JobExecution jobExecution = new JobExecution();
-        List<JobExecutionAttempt> attempts = getJobExecutionAttempts(node);
-        jobExecution.setAttempts(attempts);
+  private static final String FLOW_ID_PARAM = "flow_id";
+  private static final String FLOW_VERSION_PARAM = "flow_version";
+  private static final String PROJECT_ID_PARAM = "project_id";
+  private static final String EXPERIMENT_ID_PARAM = "experiment_id";
+
+  private static final Logger logger = LoggerFactory.getLogger(ExecutionServiceImpl.class);
+  private ExecutorManagerAdapter executorManager;
+  private ExecutionFlowDao executionFlowDao;
+
+  @Inject
+  public ExecutionServiceImpl(ExecutorManagerAdapter executorManager,
+      ExecutionFlowDao executionFlowDao) {
+    this.executorManager = executorManager;
+    this.executionFlowDao = executionFlowDao;
+  }
+
+  private String extractArgumentValue(String argName, String[] argValues) {
+    requireNonNull(argName);
+    requireNonNull(argValues);
+    if (argValues.length != 1) {
+      throw new RuntimeException(format("Argument %s should have exactly one value", argName));
+    }
+    return argValues[0];
+  }
+
+  private ExecutionBasicResponse executionResponseFromFlow(ExecutableFlow executableFlow) {
+    return new ExecutionBasicResponse(executableFlow);
+  }
+
+  @Override
+  public List<ExecutionBasicResponse> getAllExecutions(Map<String, String[]> queryParamMap) {
+    requireNonNull(queryParamMap, "query param map is null");
```

Comment:
I think we should handle these exceptions in the servlet and create appropriate responses (400 or 500).

Reply:
I am using `requireNonNull` as a check only for the situations where you would get a null pointer exception otherwise. The intention is not to use it for indicating missing values to the end user. In this case it is both checking and documenting that the method param is not expected to be null.

Reply:
Sorry, chose the wrong line to add the comment.
In line 50 above, I believe that encountering a null value there will mean the user wrote something like `/executions?**flow_id=**&project_id=1`(no value for flow_id param) which can be interpreted as a bad request especially if no default values are provided.

Reply:
Ah, I see your point. However even at line 50 above the `requireNonNull` is still only ensuring that the String array should not be null. It can be empty but that is handled by subsequent code. There is certainly more detailed param validation that can be performed, but right now the aim was to prevent the service from choking on any invalid inputs. 
(As a quick check I ran this in a debugger and and see that for args without a value are resolved to an empty string)
I do notice that the `RuntimeException` there needs to be replaced by `CloudFlowException`. I will take care of that in a subsequent PR.  
Let me know if you would like to discuss this further.

Reply:
I tested it: `req.getParameterMap()` returns an empty array in the cases when no value is provided for a query parameter. Please disregard.